include/LightGBM/sample_strategy.h

```diff
@@ -0,0 +1,45 @@
+#ifndef LIGHTGBM_SAMPLE_STRATEGY_H_
+#define LIGHTGBM_SAMPLE_STRATEGY_H_
+
+#include <LightGBM/utils/random.h>
+#include <LightGBM/utils/common.h>
+#include <LightGBM/utils/threading.h>
+#include <LightGBM/config.h>
+#include <LightGBM/dataset.h>
+#include <LightGBM/tree_learner.h>
+#include <LightGBM/objective_function.h>
+
+namespace LightGBM {
+
+class SampleStrategy {
+ public:
+  SampleStrategy() : balanced_bagging_(false), bagging_runner_(0, bagging_rand_block_) {};
+  virtual ~SampleStrategy() {};
+  static SampleStrategy* CreateSampleStrategy(const Config* config, const Dataset* train_data, const ObjectiveFunction* objective_function, int num_tree_per_iteration);
+  virtual void Bagging(int iter, TreeLearner* tree_learner, score_t* gradients, score_t* hessians) = 0;
+  virtual void Reset() = 0;
+  virtual void ResetConfig(const Config* config, bool is_change_dataset, 
+          std::vector<score_t, Common::AlignmentAllocator<score_t, kAlignedSize>>& gradients, 
+          std::vector<score_t, Common::AlignmentAllocator<score_t, kAlignedSize>>& hessians) = 0;
+  bool is_use_subset() {return is_use_subset_;}
+  data_size_t bag_data_cnt() {return bag_data_cnt_;}
```

Comment:
>bag_data_cnt [](http://example.com/codeflow?start=14&length=12)

another thing I want to mention here is the naming conversions. Looks like the function should start with the capitalized letter like Bagging, Reset, etc. So should the function bag_data_cnt be BagDataCount()? Please double check with your coach to make sure you are using the right naming conversions

Reply:
Resolved.